In [1]:
import os
from surprise import SVD, SVDpp, NMF
from surprise import Dataset, accuracy
from surprise.model_selection import cross_validate
from surprise.model_selection import GridSearchCV

# Указываем путь к файлу датасета
new_data_folder = '/data/ml-100k'
file_path = '/data/ml-100k/ml-100k/u.data'



# Загрузка датасета с помощью Surprise

data = Dataset.load_builtin("ml-100k")


In [ ]:
file_path = os.path.join(new_data_folder, 'ml-100k', 'ml-100k', 'u.data')

data = Dataset.load_builtin('ml-100k', prompt=False)


In [2]:
algo = SVD()

cross_validate(algo, data, measures=["RMSE", "MAE"], cv=5, verbose=True) 

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9328  0.9365  0.9355  0.9399  0.9331  0.9355  0.0026  
MAE (testset)     0.7356  0.7396  0.7365  0.7428  0.7333  0.7375  0.0033  
Fit time          1.02    1.02    1.14    1.29    1.12    1.12    0.10    
Test time         0.24    0.13    0.18    0.23    0.18    0.19    0.04    


{'test_rmse': array([0.93277633, 0.93646192, 0.9355238 , 0.93991636, 0.93305038]),
 'test_mae': array([0.7356359 , 0.73957162, 0.7364623 , 0.74275342, 0.73331916]),
 'fit_time': (1.0205767154693604,
  1.015610933303833,
  1.1366617679595947,
  1.287445306777954,
  1.1167919635772705),
 'test_time': (0.2422792911529541,
  0.13298892974853516,
  0.18318986892700195,
  0.23250985145568848,
  0.17563629150390625)}

Результати крос-валідації для алгоритму SVD показують наступне:
RMSE (середньо-квадратична помилка): Середнє значення: 0.9355 Стандартне відхилення: 0.0034

MAE (середня абсолютна помилка): Середнє значення: 0.7375 Стандартне відхилення: 0.0021

Час навчання (Fit time): Середнє значення: 1.00 сек Стандартне відхилення: 0.02 сек

Час тестування (Test time): Середнє значення: 0.18 сек Стандартне відхилення: 0.05 сек

Ці результати вказують на те, що модель SVD має досить низьке значення середньо-квадратичної та середньої абсолютної помилок, що вказує на гарну точність моделі при прогнозуванні рейтингів. Також, час навчання та тестування є прийнятними і не сильно відрізняються між підсумковими фолдами.

In [3]:
algo = SVDpp()

cross_validate(algo, data, measures=["RMSE", "MAE"], cv=5, verbose=True) 

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9130  0.9192  0.9171  0.9201  0.9217  0.9182  0.0030  
MAE (testset)     0.7151  0.7217  0.7207  0.7216  0.7212  0.7201  0.0025  
Fit time          16.16   15.55   16.39   15.12   16.09   15.86   0.46    
Test time         4.60    4.65    4.59    4.51    5.04    4.68    0.19    


{'test_rmse': array([0.91298791, 0.91922638, 0.91710001, 0.9200795 , 0.92167598]),
 'test_mae': array([0.71511601, 0.72165119, 0.72073323, 0.72162949, 0.72118147]),
 'fit_time': (16.160879850387573,
  15.548452138900757,
  16.394032955169678,
  15.117610931396484,
  16.091113567352295),
 'test_time': (4.597368001937866,
  4.654712438583374,
  4.5862507820129395,
  4.507249593734741,
  5.039920330047607)}

Результати крос-валідації для алгоритму SVDpp:
RMSE (середньо-квадратична помилка): Середнє значення: 0.9182 Стандартне відхилення: 0.0059

MAE (середня абсолютна помилка): Середнє значення: 0.7201 Стандартне відхилення: 0.0051

Час навчання (Fit time): Середнє значення: 15.86 сек Стандартне відхилення: 0.34 сек

Час тестування (Test time): Середнє значення: 4.68 сек Стандартне відхилення: 0.12 сек

Алгоритм SVDpp показує гарні результати з точки зору точності прогнозів, яка вимірюється RMSE та MAE. Відзначимо, що SVDpp має більший час навчання та тестування порівняно з SVD, оскільки SVDpp включає додаткові параметри та фактори, що призводить до більшої обчислювальної складності.

In [4]:
algo = NMF()

cross_validate(algo, data, measures=["RMSE", "MAE"], cv=5, verbose=True) 

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9509  0.9701  0.9633  0.9630  0.9630  0.9620  0.0062  
MAE (testset)     0.7506  0.7604  0.7573  0.7565  0.7571  0.7564  0.0032  
Fit time          2.08    2.11    2.20    2.23    2.00    2.12    0.08    
Test time         0.17    0.19    0.13    0.10    0.16    0.15    0.03    


{'test_rmse': array([0.95085747, 0.97006704, 0.96328311, 0.96302456, 0.96301582]),
 'test_mae': array([0.75058285, 0.76037754, 0.7572851 , 0.75650488, 0.75713525]),
 'fit_time': (2.0808072090148926,
  2.1080880165100098,
  2.1951098442077637,
  2.232278347015381,
  1.9970853328704834),
 'test_time': (0.17083311080932617,
  0.18758511543273926,
  0.12502026557922363,
  0.09703969955444336,
  0.15624403953552246)}

Результати крос-валідації для алгоритму NMF (Non-Negative Matrix Factorization):
RMSE (середньо-квадратична помилка): Середнє значення: 0.9620 Стандартне відхилення: 0.0058

MAE (середня абсолютна помилка): Середнє значення: 0.7564  Стандартне відхилення: 0.0062

Час навчання (Fit time): Середнє значення: 2.12 сек Стандартне відхилення: 0.03 сек

Час тестування (Test time): Середнє значення: 0.15 сек Стандартне відхилення: 0.06 сек

Алгоритм NMF також представляє певний рівень точності, але його значення RMSE та MAE трошки вищі порівняно з SVD та SVDpp. Важливо врахувати, що NMF використовує обмеження невід'ємності, що може бути корисним в деяких випадках, особливо якщо значення рейтингів не може бути від'ємним. Зате, час навчання та тестування для NMF є прийнятними.

In [5]:
#Функція для підбору найкращих параметрів за допомогою крос-валідації


def model_evaluation(algo, param_grid):
    print(f"\nAlgorithm: {algo.__name__}")
    
    grid_search = GridSearchCV(algo, param_grid, measures=['rmse', 'mae'], cv=3)
    
    grid_search.fit(data)

    print(f"Best parameters: {grid_search.best_params}")
    print(f"Best RMSE: {grid_search.best_score['rmse']:.4f}")

    algo_instance = grid_search.best_estimator['rmse']
    algo_instance.fit(data.build_full_trainset())
    predictions = algo_instance.test(data.build_full_trainset().build_testset())

    print(f"\nTrainset Accuracy (RMSE): {accuracy.rmse(predictions):.4f}")

In [6]:
param_grid = {'n_epochs': [5, 10], 'lr_all': [0.002, 0.005],
              'reg_all': [0.2, 0.4]}
algo = SVD
model_evaluation(algo, param_grid)


Algorithm: SVD
Best parameters: {'rmse': {'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.2}, 'mae': {'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.2}}
Best RMSE: 0.9557
RMSE: 0.9202

Trainset Accuracy (RMSE): 0.9202


In [7]:
param_grid = {'n_epochs': [15, 20], 'lr_all': [0.008, 0.011],
              'reg_all': [0.6, 0.8]}
algo = SVD
model_evaluation(algo, param_grid)


Algorithm: SVD
Best parameters: {'rmse': {'n_epochs': 20, 'lr_all': 0.008, 'reg_all': 0.6}, 'mae': {'n_epochs': 20, 'lr_all': 0.008, 'reg_all': 0.6}}
Best RMSE: 0.9689
RMSE: 0.9493

Trainset Accuracy (RMSE): 0.9493


Висновки:
Найкращі параметри (Перший етап):

Після першого етапу пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритSVD як:

n_epochs: 10 lr_all: 0.005 reg_all: 0.2

Найкращий RMSE на цьому етапі7 0.9550

Найкращі параметри (Другий етап):

Після другого етапу пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму SVD були змінені на:

n_epochs: 20 lr_all: 0.008 reg_all: 0.6

Найкращий RMSE на ць89на другому етапі.

Загалом, результати показують, що модель SVD показала гарні характеристики на обох етапах. Найкращі параметри можуть бути змінені під час налаштування моделі, і на цьому прикладі змінились після другого етапу. Значення RMSE на обох етапах є досить низькими, що свідчить про добру точність моделі у прогнозуванні рейтингів на всьому наборі даних. Таким чином, SVD із відповідними параметрами є ефективною моделлю для цього датасету, і вона може бути використана для рекомендаційного аналізу з задовільною точністю.ійного аналізу з задовільною точністю.довільною точністю.

In [8]:
param_grid = {'n_epochs': [5, 10], 'lr_all': [0.002, 0.005],
              'reg_all': [0.2, 0.4]}
algo = SVDpp
model_evaluation(algo, param_grid)


Algorithm: SVDpp
Best parameters: {'rmse': {'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.2}, 'mae': {'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.2}}
Best RMSE: 0.9538
RMSE: 0.9285

Trainset Accuracy (RMSE): 0.9285


In [9]:
param_grid = {'n_epochs': [15, 20], 'lr_all': [0.008, 0.011],
              'reg_all': [0.6, 0.8]}
algo = SVDpp
model_evaluation(algo, param_grid)


Algorithm: SVDpp
Best parameters: {'rmse': {'n_epochs': 20, 'lr_all': 0.008, 'reg_all': 0.6}, 'mae': {'n_epochs': 20, 'lr_all': 0.008, 'reg_all': 0.6}}
Best RMSE: 0.9684
RMSE: 0.9495

Trainset Accuracy (RMSE): 0.9495


Висновки:
Після першого етапу пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму SVD++ визначені як:

n_epochs: 10 lr_all: 0.005 reg_all: 0.2

Найкращий RMSE на цьому етапі: 0.9538

Після другого етапу пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму SVD++ були змінені на:

n_epochs: 20 lr_all: 0.008 reg_all: 0.6

Найкращий RMSE на цьому етапі: 0.9684

На обох етапах найкращі параметри були використані для тренування моделі на повному тренувальному наборі даних, і її точність була оцінена на всьому наборі даних. Оцінка моделі на весь набір даних показала RMSE рівний 0.9287 на першому етапі та 0.9495 на другому етапі. Висновки:

Загалом, результати показують, що модель SVD++ показала гарні характеристики на обох етапах. Найкращі параметри можуть бути змінені під час налаштування моделі, і на цьому прикладі змінились після другого етапу. Значення RMSE на обох етапах є досить низькими, що свідчить про добру точність моделі у прогнозуванні рейтингів на всьому наборі даних. Таким чином, SVD++ із відповідними параметрами також є ефективною моделлю для цього датасету, і вона може бути використана для рекомендаційного аналізу з задовільною точністю.

In [10]:
param_grid = {'n_factors': [10, 20, 30],
              'n_epochs': [5, 10, 20],
              'reg_pu': [0.06, 0.08, 0.1],
              'reg_qi': [0.06, 0.08, 0.1],
              'reg_bu': [0.02, 0.04, 0.6],
              'reg_bi': [0.02, 0.04, 0.6]}

algo = NMF
model_evaluation(algo, param_grid)


Algorithm: NMF
Best parameters: {'rmse': {'n_factors': 20, 'n_epochs': 20, 'reg_pu': 0.1, 'reg_qi': 0.1, 'reg_bu': 0.6, 'reg_bi': 0.6}, 'mae': {'n_factors': 20, 'n_epochs': 20, 'reg_pu': 0.1, 'reg_qi': 0.1, 'reg_bu': 0.6, 'reg_bi': 0.6}}
Best RMSE: 0.9917
RMSE: 0.9048

Trainset Accuracy (RMSE): 0.9048


Аналіз результатів:
За результатами пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму NMF визначені як: n_factors: 20 n_epochs: 20 reg_pu: 0.1 reg_qi: 0.1 reg_bu: 0.64 reg_bi: 0.62

Найкращий RMSE, отриманий за допомогою крос-валідації, становить 017909. Після отримання найкращих параметрів модель була навчена на повному тренувальному наборі даних, і її точність була оцінена на всьому наборі даних. Оцінка моделі на весь набір даних показала RMSE рівний 0.8043.

Висновки:
Отримані параметри та оцінки демонструють, що алгоритм NMF з обраною сіткою параметрів має гарні характеристики на даному наборі даних. Значення RMSE80.9043 вказує на те, що середня квадратична помилка прогнозування рейтингів на весь набір даних досить низька, що може вказувати на ефективність моделі в прогнозуванні рейтингів. Загалом, отримані результати свідчать про те, що модель NMF з вибраними параметрами є ефективною для задачі рекомендаційного аналізу на даному датасеті.